### ISSUE TITLE
Inferred type may no longer be accurate

### ISSUE THREAD

**Issue by joshuakb2** [microsoft/TypeScript#35836]
**TypeScript Version:**  3.8.0-dev.20191223

**Search Terms:** side effects, never, inferred type, inaccurate

**Code**

(Use --strictNullChecks)

```ts
class Foo {
    bar?: Foo

    public setBar(): void {
        this.bar = this;
    }

    public sayHi(): string {
        return 'Hi!';
    }
}

function test() {
    let foo = new Foo();

    foo.bar = undefined;

    foo.setBar();

    if (foo.bar) {
        // TS error: Property 'sayHi' does not exist on type 'never'.
        console.log('This should be OK: ' + foo.bar?.sayHi());
    }
}
```

**Expected behavior:** `foo.bar?.sayHi` is not an error because `foo.bar` might not be `undefined` anymore.

**Actual behavior:** TS still expects `foo.bar` to be `undefined` even though `foo.resetBar()` changes it.

**Playground Link:** [Playground](https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191221#code/MYGwhgzhAEBiD29oG8BQ0PQEZgE4H4AuORVdTABwFcsQBLYaCAUwBcAhPACgEpiA3eHQAmKcpkysAFnQgA6HLmgBeaNNkBucdAC+ZCdVoMmYAJ4AJOr2IRWuOgDsA5mIkTcbKrgfQA5JYBCXy0JPT1UADMqB2BWOngfVmZbXldMEDZoCMQVaAdmAHcSeF4tbWz4BTxc6OFmCMdmYTKJCrkWDm4eFsw6COguNsUeNLdoAHpx6AAVAGVoZlxceFxiAAVlikXWUz8IM0tfaGF4ZLz4VgWAD1lLhLVTLb98-kXfOW0JYASIeAy5EDwJxcXzTGQwCBSeBUECiLDMaAAeQA0sQjgBqLKIKoEdoHKw8braMJAA)

**Related Issues:** No.

**Comment by jcalz** [comment]
Duplicate of #9998
